Fix crash on QD when squelching CTE producer with a Motion child

l-wang opened the pull request:
Previously in commit 9fbd2da5635eecac6975e7349209373fc9177ef8, we added
special treatment for squelching share input scan by continueing
executing nodes below the CTE producer, so that a CTE consumer on a
differernt slice will not hang while waiting for the producer to finish
writing. However, this previous fix could cause QD crash if there is a
Motion node below the CTE producer, because by the time QD executes
ExecSquelchNode() in mppExecutorFinishup(), the interconnect should have
been teared down already. In this case, there is no way we could
complete executing the CTE producer subtree.

This patch added early return in execMotionUnsortedReceiver() for the
above case and removed an unnecessary assertion that would introduce
crash.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

hardikar commented on src/backend/executor/nodeMotion.c:
```diff
@@ -314,7 +315,14 @@ execMotionUnsortedReceiver(MotionState *node)
 	Assert(node->ps.state->motionlayer_context);
-	Assert(node->ps.state->interconnect_context);
+
+	if (!node->ps.state->interconnect_context && !node->stopRequested)
```
Is this not a problem for `execMotionSortedReceiver`?

l-wang replied:
Good call! It should be. Will add the same fix there.

hardikar commented on src/backend/executor/nodeShareInputScan.c:
```diff
@@ -606,6 +606,7 @@ ExecSquelchShareInputScan(ShareInputScanState *node)
 			{
 				elog(DEBUG1, "SISC WRITER (shareid=%d, slice=%d): initializing because squelched",
 					 sisc->share_id, currentSliceId);
+				estate->squelching_share_input_scan = true;
```
What is special about ShareInputScans?
I can answer my own question. It's the only node whose `ExecSquelchShareInputScan` ends up calling `ExecProcNode`, thanks 9fbd2da.

l-wang replied:
That's correct. One thing I noticed is that the test added by 9fbd2da does not serve its purpose anymore. Not only has the join order changed, even if I force the join order to be the same as described in that commit, I cannot reproduce the hang issue if I remove the special handling added by that commit. The test I added, however, will  hang without that special handling in ShareInputScan. I do plan to look into this more.

hardikar commented on src/test/regress/expected/shared_scan_optimizer.out:
```diff
@@ -0,0 +1,85 @@
+        SELECT 1 FROM cte c1, cte c2
+        )
+        FROM bar;
+ERROR:  shareinputscan with outer refs is not supported by GPDB
```
FYI this is an ERROR, I'm not sure if you wanted this.

l-wang replied:
Yes ERROR is the expectation of this test.

hardikar reviewed (commented):
Hold on, why is it legal to call `ExecSquelchMotion` (via `ExecSquelchNode` that is) when the interconnect is torn down, in general? `ExecSquelchMotion` calls `SendStopMessage`. Don't you need an interconnect to send a STOP message to the "other end"?

l-wang commented:
> Hold on, why is it legal to call `ExecSquelchMotion` (via `ExecSquelchNode` that is) when the interconnect is torn down, in general? `ExecSquelchMotion` calls `SendStopMessage`. Don't you need an interconnect to send a STOP message to the "other end"?

`ExecSquelchNode` is called at the end of multiple `Exec<Node>` functions if the node returns zero tuples. In our case, the interconnect is torn down before QD called `ExecProcNode`. `SendStopMessage` would only send the STOP message if the interconnect still exits, otherwise it only sets `pEntry->stopped` and returns.

hardikar commented on src/backend/executor/nodeShareInputScan.c:
```diff
@@ -606,6 +606,7 @@ ExecSquelchShareInputScan(ShareInputScanState *node)
 				elog(DEBUG1, "SISC WRITER (shareid=%d, slice=%d): initializing because squelched",
 					 sisc->share_id, currentSliceId);
+				estate->squelching_share_input_scan = true;
 				init_tuplestore_state(node);
```
I think we should set `squelching_share_input_scan = false` on the way out.

adam8157 commented on src/backend/executor/nodeMotion.c:
```diff
@@ -314,7 +315,14 @@ execMotionUnsortedReceiver(MotionState *node)
+
+	if (!node->ps.state->interconnect_context && !node->stopRequested)
+	{
+		if (estate->squelching_share_input_scan && Gp_role == GP_ROLE_DISPATCH)
```
I have a dumb question, could this "squelching share input scan, but continuing nodes and motion nodes below them" happen on QE?

l-wang replied:
Not a dumb question at all. It is possible for a QE to squelch a share input scan that contains a Motion node below it. However, this scenario should not happen after the interconnect is torn down on this QE, because QE would not perform `ExecSquelchNode()` after tearing down its interconnect. So, as @hardikar also pointed out, the inner if condition is really not necessary I will remove it.

adam8157 commented:
> What happens if the still executing slice is a SISC reader and the squelched slice is its SISC writer albeit with a Motion? This PR will basically stop the producer if it is squelched, since the producer depends on its descendant Motion node for tuples to complete. So if the other slow-running slice that depends on this data tries to access it later it will error or produce wrong results.

In this case, the interconnect context is not torn down yet, and this PR will not stop it, yes?

hardikar commented:
@adam8157 I see, so `TeardownInterconnect` is blocking until all the slices are cleaned up. In that case, why not `if (!node->ps.state->interconnect_context) return NULL`? Why track `estate->squelching_share_input_scan` at all?

l-wang commented:
> > What happens if the still executing slice is a SISC reader and the squelched slice is its SISC writer albeit with a Motion? This PR will basically stop the producer if it is squelched, since the producer depends on its descendant Motion node for tuples to complete. So if the other slow-running slice that depends on this data tries to access it later it will error or produce wrong results.
>> 
> In this case, the interconnect context is not torn down yet, and this PR will not stop it, yes?

@adam8157 That's correct. This PR will not return early in `execMotionUnsortedReceiver()` or `execMotionSortedReceiver()` if the interconnect is not torn down.

> @adam8157 I see, so TeardownInterconnect is blocking until all the slices are cleaned up. In that case, why not if (!node->ps.state->interconnect_context) return NULL? Why track estate->squelching_share_input_scan at all?

`TeardownInterconnect()` is executed on QD and each QE process independently. This function is only responsible for tearing down the connections related the single slice the current process is working on upon executor finish or cleanup. Your suggestion I think is valid, because after `TeardownInterconnect()` we would only call `ExecSquelchNode()`(which could lead to `ExecSquelchShareInputScan()` and then `ExecMotion()`) on the QD process if `!estate->es_got_eos`, meaning an ERROR has occurred on QD and we no longer need to fetch more tuples. I will remove the `estate->squelching_share_input_scan` flag. I only added it for safety check but now I'm convinced it is not necessary.

kainwen reviewed (commented):
This fix method seems very hacky for me. The condition here is too restrict.

And why we should not consider in QEs is not straightforward for me (although I do not build a bad case util now).

Actually, I do not like what commit https://github.com/greenplum-db/gpdb/commit/9fbd2da5635eecac6975e7349209373fc9177ef8 did.

When we are to Squelch a shared scan's writer (and we want reader not hang forever), why do we need to still need to execute plan? There are several points:

1. when squench happens, is the outerplan still executable? This pr's issue is ic is down, is anything else also down?
2. if we just want not to block readers for every, can we just return NULL instead of the hack of jesse's commit.

I did a quick and dirty test, using the following diff, it can fixes the issue in this pr also ( might not be correct, but just post my idea here for discussion ). 
 
```diff
diff --git a/src/backend/executor/nodeShareInputScan.c b/src/backend/executor/nodeShareInputScan.c
index 72dfce0d3a..c5f17c3c9f 100644
--- a/src/backend/executor/nodeShareInputScan.c
+++ b/src/backend/executor/nodeShareInputScan.c
@@ -200,7 +200,7 @@ static void ExecShareInputScanExplainEnd(PlanState *planstate, struct StringInfo
  *    is not initialized.
  */
 static void
-init_tuplestore_state(ShareInputScanState *node)
+init_tuplestore_state(ShareInputScanState *node, bool is_sq)
 {
 	EState	   *estate = node->ss.ps.state;
 	ShareInputScan *sisc = (ShareInputScan *) node->ss.ps.plan;
@@ -285,6 +285,8 @@ init_tuplestore_state(ShareInputScanState *node)

 			for (;;)
 			{
+				if (is_sq)
+					break;
 				outerslot = ExecProcNode(local_state->childState);
 				if (TupIsNull(outerslot))
 					break;
@@ -362,7 +364,7 @@ ExecShareInputScan(PlanState *pstate)

 	/* if first time call, need to initialize the tuplestore state.  */
 	if (!node->isready)
-		init_tuplestore_state(node);
+		init_tuplestore_state(node, false);

 	slot = node->ss.ps.ps_ResultTupleSlot;

@@ -542,7 +544,7 @@ ExecEndShareInputScan(ShareInputScanState *node)
 			if (currentSliceId == sisc->producer_slice_id)
 			{
 				if (!local_state->ready)
-					init_tuplestore_state(node);
+					init_tuplestore_state(node, true);
 				shareinput_writer_waitdone(node->ref, sisc->nconsumers);
 			}
 			else
@@ -581,7 +583,7 @@ ExecReScanShareInputScan(ShareInputScanState *node)
 {
 	/* On first call, initialize the tuplestore state */
 	if (!node->isready)
-		init_tuplestore_state(node);
+		init_tuplestore_state(node, true);

 	ExecClearTuple(node->ss.ps.ps_ResultTupleSlot);
 	Assert(node->ts_pos != -1);
@@ -630,7 +632,7 @@ ExecSquelchShareInputScan(ShareInputScanState *node)
 			{
 				elog(DEBUG1, "SISC WRITER (shareid=%d, slice=%d): initializing because squelched",
 					 sisc->share_id, currentSliceId);
-				init_tuplestore_state(node);
+				init_tuplestore_state(node, true);
 			}
 		}
 		else
```

l-wang commented:
Thank you @kainwen for your review! 

> This fix method seems very hacky for me. The condition here is too restrict.

I agree with you that this fix looks hacky. While I tried my best to narrow down the condition for ExecMotion to return early, I did have same concerns about potential risks. If very bad thing happens, like memory got corrupted and `interconnect_context` becomes NULL before we tear it down, early return could potentially lead to wrong results, which is much worse than crashing.

> And why we should not consider in QEs is not straightforward for me (although I do not build a bad case util now).

We did not consider QE because only QD would call `ExecSquelchNode()` after tearing down interconnect. If QE fall into this code path where it's performing ExecMotion after interconnect is down, something else must went wrong.

> Actually, I do not like what commit [9fbd2da](https://github.com/greenplum-db/gpdb/commit/9fbd2da5635eecac6975e7349209373fc9177ef8) did.
> 
> When we are to Squelch a shared scan's writer (and we want reader not hang forever), why do we need to still need to execute plan? There are several points:
> 
> 1. when squench happens, is the outerplan still executable? This pr's issue is ic is down, is anything else also down?
> 2. if we just want not to block readers for every, can we just return NULL instead of the hack of jesse's commit.
> 
> I did a quick and dirty test, using the following diff, it can fixes the issue in this pr also ( might not be correct, but just post my idea here for discussion ).

Thank you for bringing it up! I'm not a fan of commit [9fbd2da](https://github.com/greenplum-db/gpdb/commit/9fbd2da5635eecac6975e7349209373fc9177ef8) either. Let me apply your patch and explore more on this alternative solution.

kainwen commented:
> We did not consider QE because only QD would call `ExecSquelchNode()` after tearing down interconnect. If QE fall into this code path where it's performing ExecMotion after interconnect is down, something else must went wrong.

How do we verify this?

-------

The fix let motion executor just return NULL when interconnect context is gone, although some comments are there, but no guard code, maybe let some bug just skip.

kainwen commented on src/backend/executor/nodeMotion.c:
```diff
@@ -324,6 +323,21 @@ execMotionUnsortedReceiver(MotionState *node)
 
+	if (!node->ps.state->interconnect_context &&
+		Gp_role == GP_ROLE_DISPATCH && !node->ps.state->es_got_eos)
+	{
```
After talking with @l-wang , I think we can add some log here,  or might check field `es_interconnect_is_setup` in the if condition.

And the comments in https://github.com/greenplum-db/gpdb/commit/9fbd2da5635eecac6975e7349209373fc9177ef8#diff-1255433ef357570f11a775d6a32fab02584024c0708986e0b16837717ee355b0R1502 seems not correct.

Because sequence + shared scan skill using by orca, I think only case is  the whole sequence is squenched. Also I think there cannot be other slices not part of the whole sequence that use share scan reader.

But we can leave this later as a TODO.

For this pr, I think we should add some log here.

l-wang replied:
> And the comments in [9fbd2da#diff-1255433ef357570f11a775d6a32fab02584024c0708986e0b16837717ee355b0R1502](https://github.com/greenplum-db/gpdb/commit/9fbd2da5635eecac6975e7349209373fc9177ef8#diff-1255433ef357570f11a775d6a32fab02584024c0708986e0b16837717ee355b0R1502) seems not correct.
> 
> Because sequence + shared scan skill using by orca, I think only case is the whole sequence is squenched. Also I think there cannot be other slices not part of the whole sequence that use share scan reader.
> 
> But we can leave this later as a TODO.

@kainwen I just realized this comment: 

```
	 * Note that we emphatically can't "fake" an empty tuple store and just
	 * go ahead waking up the readers because that can lead to wrong results.
```

is correct! 

Here's the case:

Checkout to [this branch](https://github.com/l-wang/gpdb/commits/share-input-scan-patch). This branch is essentially the patch you posted earlier with a small fixup, and it does exactly "fake" an empty tuple store and just go ahead waking up the readers. We do really get wrong result for the following query:

```sql
CREATE TABLE foo (a int, b int);
CREATE TABLE bar (c int, d int);
CREATE TABLE jazz(e int, f int);

INSERT INTO foo values (1, 2);
INSERT INTO bar  VALUES (1, 1), (2, 2), (3, 3);
INSERT INTO jazz VALUES (2, 2), (3, 3);

ANALYZE foo;
ANALYZE bar;
ANALYZE jazz;

SET optimizer_join_order = 'query';

SELECT * FROM
        (
        WITH cte AS (SELECT * FROM foo)
        SELECT * FROM (SELECT * FROM cte UNION ALL SELECT * FROM cte)
        AS X
        JOIN bar ON b = c
        ) AS XY
        JOIN jazz on c = e AND b = f;
        

WITH cte AS (SELECT * FROM foo)
SELECT * FROM (SELECT * FROM cte UNION ALL SELECT * FROM cte)
AS X
JOIN bar ON b = c;
```

Wrong result:
```
postgres=#  SELECT * FROM
        (
        WITH cte AS (SELECT * FROM foo)
        SELECT * FROM (SELECT * FROM cte UNION ALL SELECT * FROM cte)
        AS X
        JOIN bar ON b = c
        ) AS XY
        JOIN jazz on c = e AND b = f;
 a | b | c | d | e | f 
---+---+---+---+---+---
(0 rows)
```

Expected result:
```
postgres=# RESET optimizer_join_order;
RESET
postgres=#  SELECT * FROM
        (
        WITH cte AS (SELECT * FROM foo)
        SELECT * FROM (SELECT * FROM cte UNION ALL SELECT * FROM cte)
        AS X
        JOIN bar ON b = c
        ) AS XY
        JOIN jazz on c = e AND b = f;
 a | b | c | d | e | f 
---+---+---+---+---+---
 1 | 2 | 2 | 2 | 2 | 2
 1 | 2 | 2 | 2 | 2 | 2
(2 rows)
```

Here's why:

Look at the plan:
```
                                                 QUERY PLAN                                                 
------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..2155.00 rows=1 width=24)
   ->  Hash Join  (cost=0.00..2155.00 rows=1 width=24)
         Hash Cond: (bar.c = jazz.e)
         ->  Sequence  (cost=0.00..1724.00 rows=1 width=16)
               ->  Shared Scan (share slice:id 1:0)  (cost=0.00..431.00 rows=1 width=1)
                     ->  Seq Scan on foo  (cost=0.00..431.00 rows=1 width=8)
               ->  Hash Join  (cost=0.00..1293.00 rows=1 width=16)
                     Hash Cond: (share0_ref2.b = bar.c)
                     ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..862.00 rows=1 width=8)
                           Hash Key: share0_ref2.b
                           ->  Append  (cost=0.00..862.00 rows=1 width=8)
                                 ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=1 width=8)
                                 ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=1 width=8)
                     ->  Hash  (cost=431.00..431.00 rows=1 width=8)
                           ->  Seq Scan on bar  (cost=0.00..431.00 rows=1 width=8)
         ->  Hash  (cost=431.00..431.00 rows=1 width=8)
               ->  Seq Scan on jazz  (cost=0.00..431.00 rows=1 width=8)
                     Filter: (e = f)
 Optimizer: Pivotal Optimizer (GPORCA)
(19 rows)
```

and look at the data distribution:
```
postgres=# select gp_segment_id, * from foo;
 gp_segment_id | a | b 
---------------+---+---
             1 | 1 | 2
(1 row)

postgres=# select gp_segment_id, * from bar;
 gp_segment_id | c | d 
---------------+---+---
             1 | 1 | 1
             0 | 2 | 2
             0 | 3 | 3
(3 rows)

postgres=# select gp_segment_id, * from jazz;
 gp_segment_id | e | f 
---------------+---+---
             0 | 2 | 2
             0 | 3 | 3
(2 rows)

```
On segment 1, we squelch the outer side of the Hash Join because the inner table `jazz` has 0 rows on segment 1. However, when we squelch the outer side (the Sequence node), which has the Shared Scan consumers below a Redistribute Motion, we have to really execute everything below the Motion. Because we will redistribute the tuples coming out of the Shared Scan producer to other segments. In this case, `(1,2)` from the Shared Scan (essentially table `foo`) on segment 1 will be redistributed to segment 0. Segment 0 actually do have tuples in the inner table `jazz`, so its Shared Scan won't be squelched. 

So commit 9fbd2da is legit. Thanks to @d @melanieplageman @sambitesh . I will add this test to the regression test suite.

kainwen replied:
@l-wang Thanks for the detailed explain. TIL.

kainwen reviewed (approved):
Please add some logs and consider more flag in the ifcondition.

kainwen commented:
Note, the pr pipeline fails:

```
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /tmp/build/b235ddbb/gpdb_src/src/test/regress/expected/partition_indexing.out /tmp/build/b235ddbb/gpdb_src/src/test/regress/results/partition_indexing.out
--- /tmp/build/b235ddbb/gpdb_src/src/test/regress/expected/partition_indexing.out	2021-09-13 23:04:03.668824699 +0000
+++ /tmp/build/b235ddbb/gpdb_src/src/test/regress/results/partition_indexing.out	2021-09-13 23:04:03.708828133 +0000
@@ -365,6 +365,10 @@
 drop index mpp3033a_unique1;
 drop index mpp3033a_unique2;
 drop index mpp3033a_hundred;
+DETAIL:  Process 34879 waits for AccessExclusiveLock on relation 75452 of database 16384; blocked by process 35158.
+ERROR:  deadlock detected
+Process 35158 waits for AccessShareLock on relation 75803 of database 16384; blocked by process 34879.
+HINT:  See server log for query details.
 drop index mpp3033a_stringu1;
 drop index mpp3033b_unique1;
 drop index mpp3033b_unique2;
@@ -373,6 +377,7 @@
 CREATE INDEX mpp3033a_unique1 ON mpp3033a (unique1);
 CREATE INDEX mpp3033a_unique2 ON mpp3033a (unique2);
 CREATE INDEX mpp3033a_hundred ON mpp3033a (hundred);
+ERROR:  relation "mpp3033a_hundred" already exists
 CREATE INDEX mpp3033a_stringu1 ON mpp3033a (stringu1);
 CREATE INDEX mpp3033b_unique1 ON mpp3033b (unique1);
 CREATE INDEX mpp3033b_unique2 ON mpp3033b (unique2);
```

I do not think it is related to this PR.
Also I do think the issue need to pay attention to. I create a internal story to track it.

@l-wang When you finish the code, push to re-trigger the pr pipeline.

l-wang commented:
@adam8157 @hardikar @kainwen Thank you so much for reviewing! I found all the feedback and discussions very helpful. I addressed all feedback. Please take another look if have time. I plan to merge next week.

kainwen reviewed (approved):
LGTM.

kainwen commented on src/backend/storage/file/fd.c:
```diff
@@ -2851,7 +2851,6 @@ GetTempTablespaces(Oid *tableSpaces, int numSpaces)
 {
 	int			i;
 
-	Assert(TempTablespacesAreSet());
```
Can we add a comment why we remove this?

l-wang replied:
Done.
